## Cleaning 

Importing Necessary Python Library's

In [44]:
import numpy as np
import pandas as pd

Importing Dataset :  

In [45]:
df = pd.read_csv('Unclean_Dataset.csv')

In [46]:
df.head()                           # Head() : Give Top 5 Row's as output

Unnamed: 0  Gender EthnicGroup          ParentEduc     LunchType TestPrep  \
0           0  female         NaN   bachelor's degree      standard     none   
1           1  female     group C        some college      standard      NaN   
2           2  female     group B     master's degree      standard     none   
3           3    male     group A  associate's degree  free/reduced     none   
4           4    male     group C        some college      standard     none   

  ParentMaritalStatus PracticeSport IsFirstChild  NrSiblings TransportMeans  \
0             married     regularly          yes         3.0     school_bus   
1             married     sometimes          yes         0.0            NaN   
2              single     sometimes          yes         4.0     school_bus   
3             married         never           no         1.0            NaN   
4             married     sometimes          yes         0.0     school_bus   

  WklyStudyHours  MathScore  ReadingScore  WritingScore  
0            < 5         71            71            74  
1         5 - 10         69            90            88  
2            < 5         87            93            91  
3         5 - 10         45            56            42  
4         5 - 10         76            78            75

Checking The Shape Of Dataset

In [47]:
df.shape

(30641, 15)

* Here we can see we have 30641 Row's and 15 Column's.

Get all column's Name 

In [48]:
df.columns

Index(['Unnamed: 0', 'Gender', 'EthnicGroup', 'ParentEduc', 'LunchType',
       'TestPrep', 'ParentMaritalStatus', 'PracticeSport', 'IsFirstChild',
       'NrSiblings', 'TransportMeans', 'WklyStudyHours', 'MathScore',
       'ReadingScore', 'WritingScore'],
      dtype='object')

Checking the Info about the data

In [49]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30641 entries, 0 to 30640
Data columns (total 15 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Unnamed: 0           30641 non-null  int64  
 1   Gender               30641 non-null  object 
 2   EthnicGroup          28801 non-null  object 
 3   ParentEduc           28796 non-null  object 
 4   LunchType            30641 non-null  object 
 5   TestPrep             28811 non-null  object 
 6   ParentMaritalStatus  29451 non-null  object 
 7   PracticeSport        30010 non-null  object 
 8   IsFirstChild         29737 non-null  object 
 9   NrSiblings           29069 non-null  float64
 10  TransportMeans       27507 non-null  object 
 11  WklyStudyHours       29686 non-null  object 
 12  MathScore            30641 non-null  int64  
 13  ReadingScore         30641 non-null  int64  
 14  WritingScore         30641 non-null  int64  
dtypes: float64(1), int64(4), object(10)


Checking the Statistical info about the data

In [50]:
df.describe()

Unnamed: 0    NrSiblings     MathScore  ReadingScore  WritingScore
count  30641.000000  29069.000000  30641.000000  30641.000000  30641.000000
mean     499.556607      2.145894     66.558402     69.377533     68.418622
std      288.747894      1.458242     15.361616     14.758952     15.443525
min        0.000000      0.000000      0.000000     10.000000      4.000000
25%      249.000000      1.000000     56.000000     59.000000     58.000000
50%      500.000000      2.000000     67.000000     70.000000     69.000000
75%      750.000000      3.000000     78.000000     80.000000     79.000000
max      999.000000      7.000000    100.000000    100.000000    100.000000

Checking Null Value's 

In [51]:
df.isnull()

Unnamed: 0  Gender  EthnicGroup  ParentEduc  LunchType  TestPrep  \
0           False   False         True       False      False     False   
1           False   False        False       False      False      True   
2           False   False        False       False      False     False   
3           False   False        False       False      False     False   
4           False   False        False       False      False     False   
...           ...     ...          ...         ...        ...       ...   
30636       False   False        False       False      False     False   
30637       False   False        False       False      False     False   
30638       False   False         True       False      False     False   
30639       False   False        False       False      False     False   
30640       False   False        False       False      False     False   

       ParentMaritalStatus  PracticeSport  IsFirstChild  NrSiblings  \
0                    False          False         False       False   
1                    False          False         False       False   
2                    False          False         False       False   
3                    False          False         False       False   
4                    False          False         False       False   
...                    ...            ...           ...         ...   
30636                False          False         False       False   
30637                False          False         False       False   
30638                False          False         False       False   
30639                False          False         False       False   
30640                False          False         False       False   

       TransportMeans  WklyStudyHours  MathScore  ReadingScore  WritingScore  
0               False           False      False         False         False  
1                True           False      False         False         False  
2               False           False      False         False         False  
3                True           False      False         False         False  
4               False           False      False         False         False  
...               ...             ...        ...           ...           ...  
30636           False           False      False         False         False  
30637           False           False      False         False         False  
30638           False           False      False         False         False  
30639           False           False      False         False         False  
30640           False           False      False         False         False  

[30641 rows x 15 columns]

In [52]:
df.isnull().sum()

Unnamed: 0                0
Gender                    0
EthnicGroup            1840
ParentEduc             1845
LunchType                 0
TestPrep               1830
ParentMaritalStatus    1190
PracticeSport           631
IsFirstChild            904
NrSiblings             1572
TransportMeans         3134
WklyStudyHours          955
MathScore                 0
ReadingScore              0
WritingScore              0
dtype: int64

Delete "Unnamed" column

In [53]:
df.drop(columns=["Unnamed: 0"], inplace=True)                       # "Inplace=True" For Permanent Drop

Checking Duplicated Row's

In [54]:
df.duplicated().sum()

np.int64(0)

* There are not duplicated Row's in Data. 

Checking the Data-Type

In [55]:
df.dtypes

Gender                  object
EthnicGroup             object
ParentEduc              object
LunchType               object
TestPrep                object
ParentMaritalStatus     object
PracticeSport           object
IsFirstChild            object
NrSiblings             float64
TransportMeans          object
WklyStudyHours          object
MathScore                int64
ReadingScore             int64
WritingScore             int64
dtype: object

### Filling Null Value's

In [56]:
df.isnull().sum()

Gender                    0
EthnicGroup            1840
ParentEduc             1845
LunchType                 0
TestPrep               1830
ParentMaritalStatus    1190
PracticeSport           631
IsFirstChild            904
NrSiblings             1572
TransportMeans         3134
WklyStudyHours          955
MathScore                 0
ReadingScore              0
WritingScore              0
dtype: int64

* For EthnicGroup Column:

In [57]:
df['EthnicGroup'].isnull().sum()

np.int64(1840)

In [58]:
df[df['EthnicGroup'].isnull()].sample(5)

Gender EthnicGroup          ParentEduc     LunchType   TestPrep  \
9961   female         NaN  associate's degree  free/reduced       none   
28080  female         NaN        some college      standard  completed   
14087  female         NaN                 NaN  free/reduced  completed   
8509   female         NaN   bachelor's degree      standard       none   
1255     male         NaN        some college      standard       none   

      ParentMaritalStatus PracticeSport IsFirstChild  NrSiblings  \
9961              married     sometimes          yes         1.0   
28080             married     regularly          yes         1.0   
14087             married     sometimes           no         4.0   
8509              married     sometimes          yes         1.0   
1255               single         never          yes         2.0   

      TransportMeans WklyStudyHours  MathScore  ReadingScore  WritingScore  
9961      school_bus         5 - 10         80            96            83  
28080     school_bus         5 - 10         59            60            63  
14087        private            NaN         59            79            72  
8509      school_bus         5 - 10         82            88            93  
1255      school_bus            < 5         63            58            58

In [59]:
df['EthnicGroup'].unique()

array([nan, 'group C', 'group B', 'group A', 'group D', 'group E'],
      dtype=object)

In [60]:
df['EthnicGroup'].mode()

0    group C
Name: EthnicGroup, dtype: object

Filling Null value's in "EthenicGroup" Column With it's Mode (Most Repeatitive Value)

In [61]:
df['EthnicGroup'].fillna(df['EthnicGroup'].mode()[0], inplace=True)

C:\Users\mrume\AppData\Local\Temp\ipykernel_24472\791370263.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['EthnicGroup'].fillna(df['EthnicGroup'].mode()[0], inplace=True)


In [62]:
df['EthnicGroup'].isnull().sum()

np.int64(0)

* For "ParentEduc" Column:

In [63]:
df['ParentEduc'].isnull().sum()

np.int64(1845)

In [64]:
df[df['ParentEduc'].isnull()].sample(5)

Gender EthnicGroup ParentEduc     LunchType   TestPrep  \
8354     male     group A        NaN      standard  completed   
3295   female     group B        NaN  free/reduced       none   
3642   female     group C        NaN      standard  completed   
6037     male     group C        NaN      standard       none   
21514    male     group B        NaN      standard       none   

      ParentMaritalStatus PracticeSport IsFirstChild  NrSiblings  \
8354              married     regularly          yes         3.0   
3295              married     sometimes           no         1.0   
3642              married     regularly           no         1.0   
6037              married     sometimes          yes         1.0   
21514             married     sometimes          yes         3.0   

      TransportMeans WklyStudyHours  MathScore  ReadingScore  WritingScore  
8354         private         5 - 10         97            96            96  
3295         private           > 10         43            56            58  
3642         private            NaN         62            56            66  
6037         private            < 5         70            63            57  
21514     school_bus         5 - 10         66            62            65

In [65]:
df['ParentEduc'].unique()

array(["bachelor's degree", 'some college', "master's degree",
       "associate's degree", 'high school', 'some high school', nan],
      dtype=object)

In [66]:
df['ParentEduc'].mode()

0    some college
Name: ParentEduc, dtype: object

Filling Null value's in "ParentEdue" Column With it's Mode (Most Repeatitive Value)

In [67]:
df['ParentEduc'].fillna(df['ParentEduc'].mode()[0], inplace=True)

C:\Users\mrume\AppData\Local\Temp\ipykernel_24472\785249747.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['ParentEduc'].fillna(df['ParentEduc'].mode()[0], inplace=True)


In [68]:
df['ParentEduc'].isnull().sum()

np.int64(0)

In "TestPrep" Replace "None" to "Incomplete"

In [69]:
df['TestPrep'] = df['TestPrep'].replace('none','incomplete')

Fill All The Categorical Column's With It's With The Help Of Loop

In [70]:
df.columns

Index(['Gender', 'EthnicGroup', 'ParentEduc', 'LunchType', 'TestPrep',
       'ParentMaritalStatus', 'PracticeSport', 'IsFirstChild', 'NrSiblings',
       'TransportMeans', 'WklyStudyHours', 'MathScore', 'ReadingScore',
       'WritingScore'],
      dtype='object')

In [71]:
df.dtypes

Gender                  object
EthnicGroup             object
ParentEduc              object
LunchType               object
TestPrep                object
ParentMaritalStatus     object
PracticeSport           object
IsFirstChild            object
NrSiblings             float64
TransportMeans          object
WklyStudyHours          object
MathScore                int64
ReadingScore             int64
WritingScore             int64
dtype: object

In [72]:
cat_columns = ['ParentEduc', 'LunchType', 'TestPrep', 'ParentMaritalStatus',
                'PracticeSport', 'IsFirstChild', 'TransportMeans', 'WklyStudyHours']

In [73]:
cat_columns

['ParentEduc',
 'LunchType',
 'TestPrep',
 'ParentMaritalStatus',
 'PracticeSport',
 'IsFirstChild',
 'TransportMeans',
 'WklyStudyHours']

In [74]:
for i in df.columns:
    if i in cat_columns:
        df[i].fillna(df[i].mode()[0], inplace=True)

C:\Users\mrume\AppData\Local\Temp\ipykernel_24472\1081290451.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[i].fillna(df[i].mode()[0], inplace=True)


In [75]:
df.isnull().sum()

Gender                    0
EthnicGroup               0
ParentEduc                0
LunchType                 0
TestPrep                  0
ParentMaritalStatus       0
PracticeSport             0
IsFirstChild              0
NrSiblings             1572
TransportMeans            0
WklyStudyHours            0
MathScore                 0
ReadingScore              0
WritingScore              0
dtype: int64

* For "NrSiblings" Column:

In [76]:
df['NrSiblings'].isnull().sum()

np.int64(1572)

In [77]:
df['NrSiblings'].dtype

dtype('float64')

In [78]:
df['NrSiblings'].nunique()

8

In [79]:
df['NrSiblings'].median()

np.float64(2.0)

In [80]:
df['NrSiblings'].fillna(df['NrSiblings'].median(), inplace=True)

C:\Users\mrume\AppData\Local\Temp\ipykernel_24472\4268341937.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['NrSiblings'].fillna(df['NrSiblings'].median(), inplace=True)


In [81]:
df['NrSiblings'] = df['NrSiblings'].astype(int)

In [82]:
df.isnull().sum()

Gender                 0
EthnicGroup            0
ParentEduc             0
LunchType              0
TestPrep               0
ParentMaritalStatus    0
PracticeSport          0
IsFirstChild           0
NrSiblings             0
TransportMeans         0
WklyStudyHours         0
MathScore              0
ReadingScore           0
WritingScore           0
dtype: int64

In [83]:
df.dtypes

Gender                 object
EthnicGroup            object
ParentEduc             object
LunchType              object
TestPrep               object
ParentMaritalStatus    object
PracticeSport          object
IsFirstChild           object
NrSiblings              int64
TransportMeans         object
WklyStudyHours         object
MathScore               int64
ReadingScore            int64
WritingScore            int64
dtype: object

Export Cleaned Dataset

In [84]:
df.to_csv('Cleaned_Dataset.csv')